### ISSUE TITLE
Buggy fringe display on Mac

### ISSUE THREAD

**Issue by hotpxl** [dgutov/diff-hl#49]
I'm using [Emacs for Mac OS X](http://emacsformacosx.com/) on OS X 10.10.4.

![screenshot](http://i.imgur.com/bdRS465.png)

As the screenshot shows, the fringe misses a left border.

My initialization script is like

    (require 'package)
    (package-initialize)
    (global-diff-hl-mode)

 Any ideas why?

Thanks.

**Comment by dgutov** [comment]
Is that with default Emacs configuration?

You should probably `M-x report-emacs-bug`, it's not like I can troubleshoot Mac problems, or fix its fringe rendering.

**Comment by hotpxl** [comment]
I see. I just wonder if you have any clues how to debug this. But thanks anyways.

**Comment by dgutov** [comment]
1. You haven't answered the question.

2. You could tinker with arithmetics in `diff-hl-define-bitmaps` and see if some tiny change makes the display better for you.

**Comment by hotpxl** [comment]
Oh yeah it is with default Emacs configuration.

**Comment by hotpxl** [comment]
I found the problem.  [This](https://github.com/dgutov/diff-hl/blob/master/diff-hl.el#L143) says 10 as width of the fringe. Then the next line defines a vector, expecting each line to have 10 pixels. As it turns out, I can only define vectors expecting the width of 8 pixels. Then when I'm doing `define-fringe-bitmap`, I could "stretch" it to 10 pixels.

I played with different fringe sizes, and no matter how wide my fringe is, the defined vector must be of width 8. Parts larger than 8 (or number larger than 2 ** 8 - 1) are truncated. The only way I make a wide bitmap is to define it using vector assuming width of 8, and then stretch it by passing a `w` in `define-fringe-map`.

Though this behavior is strange, I think it probably has some reason. When I print out the document for `define-fringe-map`, it says that the default value for `WIDTH` is 8. So I guess this magic number has some reason.

My suggestion is creating the vector as width 8 and stretch it. What do you think?

**Comment by dgutov** [comment]
That's the first time I'm aware that your fringe width is non-default (and above 8).

If you mean a change like this:

```diff
diff --git a/diff-hl.el b/diff-hl.el
index 17106c1..32fbb66 100644
--- a/diff-hl.el
+++ b/diff-hl.el
@@ -141,8 +141,8 @@
                    (truncate (* (frame-char-height) spacing))
                  spacing)))
          (w (frame-parameter nil (intern (format "%s-fringe" diff-hl-side))))
-         (middle (make-vector h (expt 2 (1- w))))
-         (ones (1- (expt 2 w)))
+         (middle (make-vector h (expt 2 (1- 8))))
+         (ones (1- (expt 2 8)))
          (top (copy-sequence middle))
          (bottom (copy-sequence middle))
          (single (copy-sequence middle)))
```

then it doesn't "stretch" here (on GNU/Linux), the bitmaps are just displayed as 8 pixels wide (and aligned to the right). Which seems to be the intended behavior.

FWIW, the current code performs fine for fringe widths up to 16 on GNU/Linux.

So my advice stands: report the bug for Mac OS port. Unless you have a different patch in mind.

**Comment by hotpxl** [comment]
Strangely I didn't explicitly set my fringe to be 10 pixels wide. Then I found [this](https://www.gnu.org/software/emacs/manual/html_node/elisp/Layout-Parameters.html#Layout-Parameters) on the Emacs document. Even if I set my left fringe to be 8 pixels manually, it will distribute extra width evenly and make it 10. That's the reason why default is 8 while mine is 10. I had to explicitly set it to negative 8 to make it fixed.

It still does not solve the "stretch" problem. But anyways I get a "default" uniform behavior and I am totally happy.

Sorry for taking up your time. Thanks for the help.
